### PR TITLE
Fixes empty WKT. Fixes #20753, Fixes #19190, Fixes #20754

### PR DESCRIPTION
**Pull request by lbartoletti**
## Description
This PR follows the discussion in https://github.com/qgis/QGIS/pull/9379#discussion_r269432326.

I fixed the creation of geometries when the WKT is "EMPTY" according to the standard. This corrects the 20753, 19190 and 20754 tickets.

To do this, I added a check in the parser. To make it simple and readable, in case when I find "EMPTY", I create a specific QPair.

The largest part concerns the Point which must have the possibility of also being empty. So I changed the default constructor with x and y equal to NaN. This is used to check if the point is empty. It's a matter of discussion.

I took the opportunity to correct/improve the shape classes that were supposed to use empty points but that I had modified to integrate into QGIS.

I added tests and checked locally testqgsgeometry.cpp, but I didn't run the other tests, I'll see what Travis tells me...

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [X] Commit messages are descriptive and explain the rationale for changes
- [X] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [X] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [X] New unit tests have been added for core changes
- [X] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by lbartoletti** [comment]
Exemple, with this fix:
```

for type in "POINT", "LINESTRING", "POLYGON", "GEOMETRYCOLLECTION", "MULTIPOINT", "MULTILINESTRING", "MULTIPOLYGON", "COMPOUNDCURVE", "CIRCULARSTRING", "TRIANGLE":
	for ext in "", "Z", "M", "ZM":
		g = QgsGeometry.fromWkt(type + ext + " EMPTY")
		print(g)
		print(g.asWkt())

<QgsGeometry: Point EMPTY>
Point EMPTY
<QgsGeometry: PointZ EMPTY>
PointZ EMPTY
<QgsGeometry: PointM EMPTY>
PointM EMPTY
<QgsGeometry: PointZM EMPTY>
PointZM EMPTY
<QgsGeometry: LineString EMPTY>
LineString EMPTY
<QgsGeometry: LineStringZ EMPTY>
LineStringZ EMPTY
<QgsGeometry: LineStringM EMPTY>
LineStringM EMPTY
<QgsGeometry: LineStringZM EMPTY>
LineStringZM EMPTY
<QgsGeometry: Polygon EMPTY>
Polygon EMPTY
<QgsGeometry: PolygonZ EMPTY>
PolygonZ EMPTY
<QgsGeometry: PolygonM EMPTY>
PolygonM EMPTY
<QgsGeometry: PolygonZM EMPTY>
PolygonZM EMPTY
<QgsGeometry: GeometryCollection EMPTY>
GeometryCollection EMPTY
<QgsGeometry: GeometryCollectionZ EMPTY>
GeometryCollectionZ EMPTY
<QgsGeometry: GeometryCollectionM EMPTY>
GeometryCollectionM EMPTY
<QgsGeometry: GeometryCollectionZM EMPTY>
GeometryCollectionZM EMPTY
<QgsGeometry: MultiPoint EMPTY>
MultiPoint EMPTY
<QgsGeometry: MultiPointZ EMPTY>
MultiPointZ EMPTY
<QgsGeometry: MultiPointM EMPTY>
MultiPointM EMPTY
<QgsGeometry: MultiPointZM EMPTY>
MultiPointZM EMPTY
<QgsGeometry: MultiLineString EMPTY>
MultiLineString EMPTY
<QgsGeometry: MultiLineStringZ EMPTY>
MultiLineStringZ EMPTY
<QgsGeometry: MultiLineStringM EMPTY>
MultiLineStringM EMPTY
<QgsGeometry: MultiLineStringZM EMPTY>
MultiLineStringZM EMPTY
<QgsGeometry: MultiPolygon EMPTY>
MultiPolygon EMPTY
<QgsGeometry: MultiPolygonZ EMPTY>
MultiPolygonZ EMPTY
<QgsGeometry: MultiPolygonM EMPTY>
MultiPolygonM EMPTY
<QgsGeometry: MultiPolygonZM EMPTY>
MultiPolygonZM EMPTY
<QgsGeometry: CompoundCurve EMPTY>
CompoundCurve EMPTY
<QgsGeometry: CompoundCurveZ EMPTY>
CompoundCurveZ EMPTY
<QgsGeometry: CompoundCurveM EMPTY>
CompoundCurveM EMPTY
<QgsGeometry: CompoundCurveZM EMPTY>
CompoundCurveZM EMPTY
<QgsGeometry: CircularString EMPTY>
CircularString EMPTY
<QgsGeometry: CircularStringZ EMPTY>
CircularStringZ EMPTY
<QgsGeometry: CircularStringM EMPTY>
CircularStringM EMPTY
<QgsGeometry: CircularStringZM EMPTY>
CircularStringZM EMPTY
<QgsGeometry: Triangle EMPTY>
Triangle EMPTY
<QgsGeometry: TriangleZ EMPTY>
TriangleZ EMPTY
<QgsGeometry: TriangleM EMPTY>
TriangleM EMPTY
<QgsGeometry: TriangleZM EMPTY>
TriangleZM EMPTY
```

File: src/core/geometry/qgsgeometryutils.cpp [hunk 1]
```diff
@@ -1210,12 +1210,25 @@ double QgsGeometryUtils::normalizedAngle( double angle )
...
-  QRegularExpression cooRegEx( QStringLiteral( "^[^\\(]*\\((.*)\\)[^\\)]*$" ) );
-  cooRegEx.setPatternOptions( QRegularExpression::DotMatchesEverythingOption );
-  QRegularExpressionMatch match = cooRegEx.match( wkt );
-  QString contents = match.hasMatch() ? match.captured( 1 ) : QString();
+  QString wktParsed = wkt;
+  QString contents;
+  if ( wkt.contains( QString( "EMPTY" ), Qt::CaseInsensitive ) )
+  {
+    QStringList parts = wkt.split( " " );
+    if ( ( parts.count() == 2 ) && ( parts[1].toUpper() == QString( "EMPTY" ) ) )
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeometryutils.cpp]:
```suggestion
    if ( ( parts.count() == 2 ) && ( parts[1].compare( QLatin1String( "EMPTY" ), Qt::CaseInsensitive ) == 0 ) )
```

File: src/core/geometry/qgslinestring.cpp [hunk 2]
```diff
@@ -375,9 +377,15 @@ QByteArray QgsLineString::asWkb() const
 QString QgsLineString::asWkt( int precision ) const
 {
   QString wkt = wktTypeStr() + ' ';
-  QgsPointSequence pts;
-  points( pts );
-  wkt += QgsGeometryUtils::pointsToWKT( pts, precision, is3D(), isMeasure() );
+
+  if ( isEmpty() )
+    wkt += "EMPTY";
```
Comment (nyalldawson) [inline src/core/geometry/qgslinestring.cpp]:
```suggestion
    wkt += QStringLiteral( "EMPTY" );
```

File: src/core/geometry/qgspoint.cpp [hunk 3]
```diff
@@ -222,13 +225,20 @@ QByteArray QgsPoint::asWkb() const
...
-  wkt += qgsDoubleToString( mX, precision ) + ' ' + qgsDoubleToString( mY, precision );
-  if ( is3D() )
-    wkt += ' ' + qgsDoubleToString( mZ, precision );
-  if ( isMeasure() )
-    wkt += ' ' + qgsDoubleToString( mM, precision );
-  wkt += ')';
+  QString wkt = wktTypeStr();
+
+  if ( isEmpty() )
+    wkt += QString( " EMPTY" );
```
Comment (nyalldawson) [inline src/core/geometry/qgspoint.cpp]:
```suggestion
    wkt += QStringLiteral( " EMPTY" );
```

**Comment by 3nids** [comment]
This might break a few plugins, no? 
Will this be backported to 3.4? I'd say ithe point part (NaN) needs so we have an homogeneous behavior for plugins.

**Comment by nyalldawson** [comment]
> This might break a few plugins, no?
Will this be backported to 3.4? I'd say ithe point part (NaN) needs so we have an homogeneous behavior for plugins.

I was wondering that too - but I actually don't think the change is as scary as it looks at first. If you don't use an empty point, you won't get one.

One thing I'm not sure of is how this needs to be handled by providers. I suspect the wkb encoding/decoding may also need upgrading to handle empty points. And then individual providers will also need updates to be able to store and read empty points correctly.

**Comment by lbartoletti** [comment]
@3nids 
> This might break a few plugins, no?

@nyalldawson 
> I was wondering that too - but I actually don't think the change is as scary as it looks at first. If you don't use an empty point, you won't get one.

I had the same thought. I agree with you.
It was a test after our discussion with Matthias and it was less painful than I expected.

> One thing I'm not sure of is how this needs to be handled by providers. I suspect the wkb encoding/decoding may also need upgrading to handle empty points. And then individual providers will also need updates to be able to store and read empty points correctly.

Yes, and same for JSON, GML, etc.

File: src/core/geometry/qgsgeometryutils.cpp [hunk 4]
```diff
@@ -1210,12 +1210,25 @@ double QgsGeometryUtils::normalizedAngle( double angle )
...
-
-  QRegularExpression cooRegEx( QStringLiteral( "^[^\\(]*\\((.*)\\)[^\\)]*$" ) );
-  cooRegEx.setPatternOptions( QRegularExpression::DotMatchesEverythingOption );
-  QRegularExpressionMatch match = cooRegEx.match( wkt );
-  QString contents = match.hasMatch() ? match.captured( 1 ) : QString();
+  QString wktParsed = wkt;
+  QString contents;
+  if ( wkt.contains( QString( "EMPTY" ), Qt::CaseInsensitive ) )
+  {
+    QStringList parts = wkt.split( " " );
```
[the remark this answers is not in the document]
Comment (lbartoletti) [reply src/core/geometry/qgsgeometryutils.cpp]: Done with a RegEx allowing "<spaces>WKBTYPE<spaces>EMPTY<spaces>"

File: src/core/geometry/qgscircularstring.cpp [hunk 5]
```diff
@@ -311,6 +311,9 @@ bool QgsCircularString::fromWkt( const QString &wkt )
     return false;
   mWkbType = parts.first;
 
+  if ( parts.second == QString( "EMPTY" ) )
```
[the remark this answers is not in the document]
Comment (lbartoletti) [reply src/core/geometry/qgscircularstring.cpp]: `wktReadBlock` will always returns "EMPTY", so it's not necessary to check the sensitivity of the case.

I added random tests.

**Comment by lbartoletti** [comment]
@3nids I have a problem with SIP introduced by changing the QgsPoint constructor:

```
>>> pt = QgsPointXY(1,2)
>>> QgsPoint(pt)
TypeError: must be real number, not QgsPointXY

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/code.py", line 91, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
SystemError: <class 'qgis._core.QgsPoint'> returned a result with an error set
```

Any idea?

**Review by 3nids — APPROVED** [review]
All good for me for the invalid point.

**Comment by lbartoletti** [comment]
> All good for me for the invalid point.

Even if it is possible to test if a point `== QgsPoint()`, I think a good practice now is to test a point by `QgsPoint.isEmpty()`

The new constructor breaks the possibility of doing a `QgsPoint( QgsPointXY())`
Do I have to add a line in the dox?

**Comment by nyalldawson** [comment]
I wonder if we should take the same time to unify the behavior with QgsPointXY - that also currently treats a null point and a valid (0, 0) point identically.

**Comment by lbartoletti** [comment]
> I wonder if we should take the same time to unify the behavior with QgsPointXY - that also currently treats a null point and a valid (0, 0) point identically.

You're right, I'll take care of it.

**Comment by nyalldawson** [comment]
Fantastic -- thank you!

**Comment by lbartoletti** [comment]
@nyalldawson Done. I also added new tests for QgsPointXY.
Travis errors (Tiemout) are not related to my tests.

**Comment by lbartoletti** [comment]
@nyalldawson If it's good for you, I'll let you merge it. I'm going to tackle JSON now :wink:

**Comment by lbartoletti** [comment]
@nyalldawson @3nids ready to merge?

**Comment by luipir** [comment]
> @nyalldawson @3nids ready to merge?

failing tests are related to this PR?

**Comment by luipir** [comment]
> seems timeout only! restarted build

seems same fails... @lbartoletti can you check if something related with  your pr? what about your local tests?

**Comment by lbartoletti** [comment]
@luipir Hmm... you're right, there is something with GEOS related with an empty point somewhere.

**Comment by lbartoletti** [comment]
please, not now stale bot!

File: src/core/geometry/qgscircularstring.cpp [hunk 6]
```diff
@@ -311,6 +311,9 @@ bool QgsCircularString::fromWkt( const QString &wkt )
     return false;
   mWkbType = parts.first;
 
+  if ( parts.second.compare( QLatin1String( "EMPTY" ) ) == 0 )
```
Comment (nyalldawson) [inline src/core/geometry/qgscircularstring.cpp]:
```suggestion
  if ( parts.second.compare( QLatin1String( "EMPTY" ), Qt::CaseInsensitive ) == 0 )
```

File: src/core/geometry/qgscompoundcurve.cpp [hunk 7]
```diff
@@ -176,6 +176,9 @@ bool QgsCompoundCurve::fromWkt( const QString &wkt )
     return false;
   mWkbType = parts.first;
 
+  if ( parts.second.compare( QLatin1String( "EMPTY" ) ) == 0 )
```
Comment (nyalldawson) [inline src/core/geometry/qgscompoundcurve.cpp]: Case insensitive, etc

File: src/core/geometry/qgspoint.cpp [hunk 8]
```diff
@@ -674,7 +684,7 @@ QgsPoint QgsPoint::project( double distance, double azimuth, double inclination
 
 bool QgsPoint::isEmpty() const
 {
-  return false;
+  return std::isnan( mX ) || std::isnan( mY );
```
Comment (nyalldawson) [inline src/core/geometry/qgspoint.cpp]: QgsGeometry::asPoint SIP code will need updating to raise an exception if an empty point is attempted to convert to a QgsPointXY

File: src/core/pal/pointset.cpp [hunk 9]
```diff
@@ -135,15 +135,17 @@ void PointSet::createGeosGeom() const
   GEOSCoordSequence *coord = GEOSCoordSeq_create_r( geosctxt, nbPoints + ( needClose ? 1 : 0 ), 2 );
   for ( int i = 0; i < nbPoints; ++i )
   {
-    GEOSCoordSeq_setX_r( geosctxt, coord, i, x[i] );
-    GEOSCoordSeq_setY_r( geosctxt, coord, i, y[i] );
+    // GEOS doesn't supports nan coordinates
+    GEOSCoordSeq_setX_r( geosctxt, coord, i, std::isnan( x[i] ) ? 0 : x[i] );
```
Comment (nyalldawson) [inline src/core/pal/pointset.cpp]: Shouldn't we just skip these? (Although I doubt a nan vertex will reach this far anyway)

File: src/core/qgspointxy.h [hunk 10]
```diff
@@ -147,6 +147,9 @@ class CORE_EXPORT QgsPointXY
      */
     QPointF toQPointF() const
     {
+      if ( isEmpty() ) // special case
+        return QPointF();
```
Comment (nyalldawson) [inline src/core/qgspointxy.h]: Note that a null QPointF is actually QPointF( 0, 0 ). But I can't see how else we could handle this.

File: tests/src/python/test_qgsgeometry.py [hunk 11]
```diff
@@ -1305,7 +1305,9 @@ def testVertexAt(self):
         polyline = QgsGeometry.fromPolylineXY(points)
 
         for i in range(0, len(points)):
-            self.assertEqual(QgsPoint(points[i]), polyline.vertexAt(i), "Mismatch at %d" % i)
+            # WORKAROUND to avoid a system error
```
Comment (nyalldawson) [inline tests/src/python/test_qgsgeometry.py]: This looks scary -- what happens here?

**Review by nyalldawson — COMMENTED** [review]
I think we may also need to handle empty points in the geometry vertex iterator -- e.g.

    for p in QgsGeometry.fromWkt('Point empty').vertices():
        assert False

should never loop.

Otherwise, this is looking good. I think we should defer merging till after 3.8.0 -- that'll give a full dev cycle to iron out any issues.

**Comment by lbartoletti** [comment]
> I think we may also need to handle empty points in the geometry vertex iterator -- e.g.
> 
> ```
> for p in QgsGeometry.fromWkt('Point empty').vertices():
>     assert False
> ```
> 
> should never loop.

True. 
I'll take a look at this.

> Otherwise, this is looking good. 

Not totally. I still have an issue for a test with QgsPointXY. Last weeks I didn't have time to explore this issue. But, I think that is maybe not a good option to add default value to NaN for QgsPointXY.
This class is used for the screen, as QPointF for example, it could be (0,0) by default. So, how to know if it's empty. Maybe by adding a member `mIsEmpty` setted to false only with the default constructor and changed to true when `set`, `setX` or `setY` are called?

> I think we should defer merging till after 3.8.0 -- that'll give a full dev cycle to iron out any issues.

I agree with you.

Thanks

File: src/core/pal/pointset.cpp (hunk 9 above)
Comment (lbartoletti) [reply src/core/pal/pointset.cpp]: Done

File: tests/src/python/test_qgsgeometry.py (hunk 11 above)
Comment (lbartoletti) [reply tests/src/python/test_qgsgeometry.py]: I'll retry but it's not the first time I have encountered this kind of problem. Maybe a problem with SIP?

Comment (lbartoletti) [reply tests/src/python/test_qgsgeometry.py]: @nyalldawson OK, it's come from the newest signature of QgsPoint.
My previous comment: https://github.com/qgis/QGIS/pull/9645#issuecomment-477904611

File: src/core/geometry/qgspoint.cpp (hunk 8 above)
Comment (lbartoletti) [reply src/core/geometry/qgspoint.cpp]: @nyalldawson Why?
An empty QgsPoint could be converted as an empty QgsPointXY, https://github.com/qgis/QGIS/pull/9645/files#diff-79b56e84ace3386453d7ffde919a808bR35

**Comment by lbartoletti** [comment]
@nyalldawson A summary of all changes made by this PR:

- All geometric types have an EMPTY definition with a WKT format compliant to the OGC standard
- Especially QgsPoint have a new signature and the default constructor return an empty point. There is a change in python since it is no longer possible to do `QgsPoint (QgsPointXY())` for example.
- QgsPointXY have also an isEmpty method but since this class is mainly used on the canvas, I didn't change x and y by default to NaN instead of 0. I came back to a behavior close to the point classes used for the screens like QPointF and not the geometry. This would have made too many changes in the API. I think it's something to discuss for QGIS 4.
- I have strengthened the comparison operators to support NaN
- and I added tests for all the modifications

**Comment by lbartoletti** [comment]
@nyalldawson Do you think you can take a look at it and merge it before QGIS 3.10?

**Comment by lbartoletti** [comment]
Up :wink:

**Comment by lbartoletti** [comment]
@nyalldawson @luipir @3nids Someone to press the green button? :wink:

**Pull request merged**